flink/src/main/java/org/apache/iceberg/flink/source/assigner/SplitAssigner.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink.source.assigner;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import javax.annotation.Nullable;
+import org.apache.flink.api.connector.source.SplitEnumeratorContext;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplitStatus;
+
+/**
+ * Enumerator should call the assigner APIs from the coordinator thread.
+ * This is to simplify the thread safety for assigner implementation.
+ */
+public interface SplitAssigner extends AutoCloseable {
+
+  /**
+   * Some assigners may need to start background threads or perform other activity such as
+   * registering as listeners to updates from other event sources e.g., watermark tracker.
+   */
+  default void start() {
+  }
+
+  /**
+   * Some assigners may need to perform certain actions
+   * when their corresponding enumerators are closed
+   */
+  @Override
+  default void close() {
+  }
+
+  /**
+   * Request a new split from the assigner
+   * as enumerator trying to assign splits to awaiting readers
+   */
+  GetSplitResult getNext(@Nullable String hostname);
```

Comment:
Add a comment here saying that this function assumes that the split returned for the host will be assigned by the SplitEnumerator. otherwise, this could lead to an inconsistent state as the assigner would assume the split was assigned but the reader may not have received it.

Reply:
that is true. but I am not sure we need to document the enumerator behavior in the assigner interface/contract. Assigner is responsible of returning a split if available.

Reply:
Yes but it also changes its internal state right? In other words, this method has a side-effect even though it's just a getter. Normally, getters are meant to be side-effect free. Thus, it might make sense to document that the assigner would assume the returned split will be assigned to some subtask; otherwise, it might lead to an inconsistent state between the enumerator & the assigner.

Reply:
side effect is a good point. Maybe we should also rename the method to `assignSplit`. sure, we can document the expectation as you described